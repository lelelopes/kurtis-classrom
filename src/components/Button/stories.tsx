import { ComponentStory, ComponentMeta } from '@storybook/react'
import { RightArrowAlt } from '@styled-icons/boxicons-regular/RightArrowAlt'

import { Button } from '.'

export default {
  title: 'Button',
  component: Button,
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/oDgQyWOou0beMgQTbPYblY/Landing-page-for-educational-app-(Community)?node-id=303%3A113',
    },
  },
  args: {
    children: 'Button',
    size: 'medium',
    variant: 'primary',
  },
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => {
  return (
    <div>
      <Button {...args} />
    </div>
  )
}

export const Default = Template.bind({})

export const Secondary = Template.bind({})
Secondary.args = {
  variant: 'secondary',
  size: 'small',
}

export const Outline = Template.bind({})
Outline.args = {
  variant: 'outline',
  size: 'small',
}

export const WithIcon = Template.bind({})
WithIcon.args = {
  size: 'small',
  children: 'Write the code here',
  icon: <RightArrowAlt />,
}

WithIcon.argTypes = {
  size: {
    type: 'string',
  },
}

WithIcon.parameters = {
  backgrounds: {
    default: 'dark',
  },
}
