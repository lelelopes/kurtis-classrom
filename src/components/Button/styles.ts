import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'

import { ButtonProps } from '.'

export type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'variant'>

const wrapperModifiers = {
  primary: (theme: DefaultTheme) => css`
    transition: background ${theme.transition.default};
    background: ${theme.colors.primary};
  `,
  secondary: (theme: DefaultTheme) => css`
    box-shadow: none;
    transition: background ${theme.transition.default};
    background: ${theme.colors.secondary};

    &:hover {
      background: ${darken(0.1, theme.colors.secondary)};
    }
  `,
  outline: (theme: DefaultTheme) => css`
    box-shadow: none;
    background: none;
    border: 1px solid ${theme.colors.gray};
    color: ${theme.colors.gray};
    transition: filter ${theme.transition.default};

    &:hover {
      filter: brightness(0.8);
    }
  `,
  small: () => css`
    width: 11rem;
    height: 4.4rem;
  `,
  medium: () => css`
    width: 15.9rem;
    height: 5.3rem;
  `,
  large: () => css`
    width: 25.5rem;
    height: 5.6rem;
  `,
  withIcon: (theme: DefaultTheme) => css`
    width: 33rem;
    height: 7.3rem;
    background: ${theme.colors.white};
    border-radius: ${theme.border.radius.huge};
    color: ${theme.colors.gray};
    font-size: ${theme.font.sizes.xlarge};
    box-shadow: none;

    &:hover {
      filter: brightness(0.95);
    }

    span {
      font-size: ${theme.font.sizes.xlarge};
      line-height: ${theme.font.sizes.xlarge};
      font-style: normal;
      font-weight: ${theme.font.bold};
      padding: 2.4rem;

      & + svg {
        width: 5.2rem;
      }
    }
  `,
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, variant, hasIcon }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.extraBold};
    border: 0;
    border-radius: ${theme.border.radius.small};
    cursor: pointer;
    text-decoration: none;
    box-shadow: inset 0px -4px 1px rgba(0, 0, 0, 0.6);
    transition: background ${theme.transition.fast};

    &:hover {
      background: ${`${
        hasIcon || variant === 'outline' ? '' : theme.colors.hover
      }`};
    }

    ${!!size && wrapperModifiers[size]()};
    ${variant === 'primary' && wrapperModifiers.primary(theme)};
    ${variant === 'secondary' && wrapperModifiers.secondary(theme)};
    ${variant === 'outline' && wrapperModifiers.outline(theme)};
    ${hasIcon && wrapperModifiers.withIcon(theme)};
  `}
`
