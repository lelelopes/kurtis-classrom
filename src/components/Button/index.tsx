import { ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes = ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  variant?: 'primary' | 'secondary' | 'outline'
  icon?: JSX.Element
} & ButtonTypes

export const Button = ({
  children,
  variant = 'primary',
  size = 'small',
  icon,
  ...props
}: ButtonProps) => (
  <S.Wrapper variant={variant} size={size} hasIcon={!!icon} {...props}>
    {!!children && <span>{children}</span>}
    {icon}
  </S.Wrapper>
)

export default Button
