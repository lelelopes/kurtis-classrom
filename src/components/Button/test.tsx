import { screen } from '@testing-library/react'
import { darken } from 'polished'
import { RightArrowAlt } from '@styled-icons/boxicons-regular/RightArrowAlt'

import { renderWithTheme } from 'utils/test/helpers'

import theme from 'styles/themes/default'

import Button from '.'
import { ThemeProvider } from 'styled-components'

const { getByRole, getByText, getByTestId } = screen

describe('<Button />', () => {
  it('should render by default', () => {
    const { container } = renderWithTheme(<Button>Button</Button>)

    const button = getByRole('button', { name: 'Button' })

    expect(button).toHaveStyle({
      background: theme.colors.primary,
      width: '11rem',
      height: '4.4rem',
      'font-size': theme.font.sizes.medium,
      'border-radius': theme.border.radius.small,
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with the correct sizes', () => {
    const { rerender } = renderWithTheme(<Button size="medium">Button</Button>)

    const button = getByRole('button', { name: 'Button' })

    expect(button).toHaveStyle({
      width: '15.9rem',
      height: '5.3rem',
    })

    rerender(
      <ThemeProvider theme={theme}>
        <Button size="large">Button</Button>
      </ThemeProvider>,
    )

    expect(button).toHaveStyle({
      width: '25.5rem',
      height: '5.6rem',
    })
  })

  it('should render the secondary variant', () => {
    renderWithTheme(<Button variant="secondary">Button</Button>)

    const button = getByRole('button', { name: 'Button' })

    expect(button).toHaveStyle({
      background: theme.colors.secondary,
      'box-shadow': 'none',
    })

    expect(button).toHaveStyleRule(
      'background',
      darken(0.1, theme.colors.secondary),
      {
        modifier: ':hover',
      },
    )
  })

  it('should render a icon version', () => {
    renderWithTheme(
      <Button icon={<RightArrowAlt data-testid="icon" />}>Button</Button>,
    )

    const span = getByText('Button')
    const icon = getByTestId('icon')
    const button = getByRole('button', { name: 'Button' })

    expect(span).toBeInTheDocument()
    expect(span).toHaveStyle({
      'font-size': theme.font.sizes.xlarge,
      'line-height': theme.font.sizes.xlarge,
      'font-style': 'normal',
      'font-weight': theme.font.bold,
      padding: '2.4rem',
    })

    expect(icon).toBeInTheDocument()

    expect(button).toHaveStyle({
      width: '33rem',
      height: '7.3rem',
      'box-shadow': 'none',
      background: theme.colors.white,
      'border-radius': theme.border.radius.huge,
      color: theme.colors.gray,
      'font-size': theme.font.sizes.xlarge,
    })
  })

  it('should render the outline variant', () => {
    renderWithTheme(<Button variant="outline">Button</Button>)

    const button = getByRole('button', { name: 'Button' })

    expect(button).toHaveStyle({
      'box-shadow': 'none',
      background: 'none',
      border: `1px solid ${theme.colors.gray}`,
      color: theme.colors.gray,
    })
  })
})
