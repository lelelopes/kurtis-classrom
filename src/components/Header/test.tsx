import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import Header from '.'

const { getByText, getByRole, getByLabelText } = screen

describe('<Header />', () => {
  it('should render correctly Header', () => {
    renderWithTheme(<Header />)

    expect(getByText(/en/i)).toBeInTheDocument()
    expect(getByRole('button', { name: /sign up/i })).toBeInTheDocument()
    expect(getByRole('button', { name: /log in/i })).toBeInTheDocument()
    expect(getByLabelText(/abrir menu/i)).toBeInTheDocument()
  })
})
