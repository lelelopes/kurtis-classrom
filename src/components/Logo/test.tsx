import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test/helpers'

import theme from 'styles/themes/default'

import Logo from '.'

describe('<Logo />', () => {
  it('should render by default', () => {
    const { container } = renderWithTheme(<Logo />)

    const icon = container.querySelector('svg > path:nth-child(1)')
    const text = container.querySelector('svg > path:nth-child(2)')

    expect(screen.getByLabelText('Kurtis Classrom').parentElement).toHaveStyle({
      width: '41.3rem',
      height: '5.1rem',
    })

    expect(icon).toHaveStyle({
      color: theme.colors.primary,
    })

    expect(text).toHaveStyle({
      color: theme.colors.dark,
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render on secondary version', () => {
    const { container } = renderWithTheme(<Logo color="secondary" />)

    const icon = container.querySelector('svg > path:nth-child(1)')
    const text = container.querySelector('svg > path:nth-child(2)')

    expect(icon).toHaveStyle({
      color: theme.colors.white,
    })

    expect(text).toHaveStyle({
      color: theme.colors.white,
    })
  })

  it('should render the Logo in small size', () => {
    renderWithTheme(<Logo size="small" />)

    expect(screen.getByLabelText('Kurtis Classrom').parentElement).toHaveStyle({
      width: '21.8rem',
      height: '2.8rem',
    })
  })

  it('should render the Logo in medium size', () => {
    renderWithTheme(<Logo size="medium" />)

    expect(screen.getByLabelText('Kurtis Classrom').parentElement).toHaveStyle({
      width: '37.2rem',
      height: '4.8rem',
    })
  })
})
