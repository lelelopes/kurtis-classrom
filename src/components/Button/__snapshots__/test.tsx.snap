// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FFFFFF;
  font-size: 1.6rem;
  font-weight: 800;
  border: 0;
  border-radius: 0.4rem;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: inset 0px -4px 1px rgba(0,0,0,0.6);
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
  width: 11rem;
  height: 4.4rem;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  background: #3A5AFF;
}

.c0:hover {
  background: #FFC43B;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`
